meson.build

```diff
@@ -252,6 +252,21 @@ if host_machine.system() in ['windows', 'cygwin']
   winmm_dep = cxx.find_library('winmm', required : true)
 endif
 
+# For GNU extensions, we need to define _GNU_SOURCE and tell Meson to
+# compile and link with thread_dep dependency when checking this feature.
+if cc.links('''
+            #ifndef _GNU_SOURCE
+            #define _GNU_SOURCE
+            #endif
+            #include <pthread.h>
+            int main() {
+                return pthread_setname_np(pthread_self(), "test");
+            }
+            ''',
+            name : 'pthread_setname_np',
+            dependencies : threads_dep)
+    conf_data.set10('HAVE_PTHREAD_SETNAME_NP', true)
+endif
```

Comment:
> `dependencies : dependency('threads'))`

:arrow_up: That's what I was missing when I originally tried:
   `cc.has_function('pthread_setname_np', '#include <pthread.h>')`

Can we be more specific about the function we test for?

On macOS, `pthread.h` only supports setting the name when inside the thread:

``` c++
int pthread_setname_np(const char*);
```

I'm worried this check will return true in this case, but it's not the function we want.

Reply:
I've been combing the meson documentation but can't see any other way (besides running a link test) to check for the presence of a function _and_ its arguments_ and types.

So without a link-check, we get this on macOS:

![2021-03-12_09-27](https://user-images.githubusercontent.com/1557255/110976025-40399700-8315-11eb-98ff-03f4e3e76701.png)

Should I exclude `MACOSX` on the source-side?  

``` c++
#if HAVE_PTHREAD_SETNAME_NP && !defined(MACOSX)
```

Reply:
**Edit** instead of excluding MacOS, I opted to turn this into a positive check for GNU extensions, which we also use in other support functions.

``` c++
#if defined(HAVE_PTHREAD_SETNAME_NP) && defined(_GNU_SOURCE)
```